Updated style guide to match the latest best practices (#2)
ckeditor/ckeditor-idiomatic.js: readme.md

# Code style guide for [CKEditor 4](https://ckeditor.com/ckeditor/ckeditor4)

**Table of contents**

> * [2. Code style guide](#2-code-style-guide)

Reviewer: Starting numeration from `2` looks not natural. I feel it is better to add at least 1. Introduction before it, for not confusing anyone.

Author: I will add missing headers instead, so numbering is consistent with headers.

		bar: 'bar'
	};

```

##### 2.2.2.2. One variable per scope

Using only one variable per scope (function) promotes readability and keeps your declaration list free of clutter (also saves a few keystrokes). However, if a scope have [early returns](281-early-returns) statements, it will be more readable and performance wise declaring variable later, when definition is needed.

```javascript
// Bad.
function foo() {
	// 2 statements

	var bar = '';

	// 2 statements

	var foo = '';

	// 10 statements

	var bom = '';

	// some statements
}

// Good
function foo() {
	var bar = '',
		foo = '';

	// some statements
}

// Bad
function bar( x ) {
	var foo = doSomethingSlow();

	if ( x < 2 ) {
		return;
	}

	// some statements
}

// Good
function bar( x ) {
	if ( x < 2 ) {
		return;
	}

	var foo = doSomethingSlow();

	// some statements
}
```

##### 2.2.2.3. Function declaration

Prefer function declaration instead of function expression to avoid issues with executing function before it has been defined.

```javascript
// Good
function square( number ) {
	return number * number;
}

square( 10 );

// Bad
var square = function( number ) {
	return number * number;
};

square( 10 );

> ```

Reviewer: Shouldn't be here also visible case related to: "to avoid issues with executing function before it has been defined"
So code should look like:
```javascript
// Good
square( 10 );

function square( number ) {
	return number * number;
}

// Bad
square( 10 );

var square = function( number ) {
	return number * number;
};
```

##### 2.2.2.4. Continuation-Passing style

> If your function should be executed asynchronously or requires additional callback (e.g. filter function) as an function argument, pass it as the last argument in function signature. Prefer [Promises](#2227-promises) for asynchronous code when possible.

Reviewer: Link is not working

```javascript
// Declaration.
function square( number, callback ) {
	callback( number * number );
}

// Usage.
square( 10, function( square ) {
	// callback statements
} );
```

##### 2.2.2.5. Promises

Preferably, asynchronous code is written using Promises. It's not always possible when maintaining code, however, prefer Promise based code for new features and everythere where you have control over API structure consumer. Always use `CKEDITOR.tools.promise` instead of native `Promise` object for wider browser support.

```javascript
var promise = new CKEDITOR.tools.promise( function( resolve, reject ) {
	var result = getContentFromFileServer( options );

	if ( result.status === 200 ) {
		resolve( result.data );
	} else {
		reject( result.error );
	}
} );

> ```

Reviewer: This example is synchronous code and doesn't show exact how to use promises. WDYT about slight modification:
```javascript
var promise = new CKEDITOR.tools.promise( function( resolve, reject ) {
    getContentFromFileServer( options, function( result ) {
        if ( result.status === 200 ) {
            resolve( result.data );
        } else {
            reject( result.error );
        }
    } );
} );
```

Author: Huh, funny, I obviously had async example in mind but looks like I rushed it a bit. It should look like your listing.

##### 2.2.2.6 Constructor declaration

Prefer using [CKEDITOR.tools.createClass](https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_tools.html#method-createClass) API instead of native constructor declaration.

```javascript
// Good
var FooBar = CKEDITOR.tools.createClass( {
	$: function( message ) {
		this.message = message;
	},
	
	proto: {
		greeting: function() {
			console.log( this.message );
		}

> 	}

Reviewer: `createClass` supports also privates, inheritance and static methods. Maybe we could show an example for it as well. It doesn't have to be the same example, there could be multiple examples or code references showing different usages of this tool.
https://github.com/ckeditor/ckeditor4/blob/411985373e833a0df76e4ebd0afb45f337612725/core/tools.js#L759-L763

Author: I'm not sure about that. The main information seems to be included i.e. use `createClass` with API reference, do we need copy docs here?

typeof variable === 'object'
```

Array:

```javascript
// In CKEditor:
CKEDITOR.tools.isArray( arrayLikeObject )
// In other cases:
Array.isArray( arrayLikeObject )
(wherever possible)
```

Node:

```javascript
// In CKEditor:
elem.type === CKEDITOR.NODE_ELEMENT
// In other cases:
elem.nodeType === Node.ELEMENT_NODE
```

null:

```javascript
variable === null
```

undefined:

```javascript
variable === undefined
```

#### 2.3.2. Coerced Types

Use coersion with caution. In most cases prefer explicit type casting than using implicit operator coersion. Note that some practices are more common than the others, so use common sense to choose correct casting method.

##### 2.3.2.1. Boolean coersion

Boolean coersion leveraged by double `!!` is widely adopted in CKEditor 4 codebase and it's one of the exception where you should use implicit coersion to its popularity.

Small explanation how double `!!` works:

1. The first `!` negates value and parses it into boolean value.
2. The second `!` flips negation.

```javascript
var string = 'some string value',
	object = {
		foo: 'foo'
	},
	number = 0,
	nothing = null;

!!string
// true

!!number
// false

!!object
// true

!!nothing

> // false

Reviewer: I would also add information about preserving extra caution when coersion is used with comparing signs.
```javascript
// Please preserve extra caution when coerse Array or Objects and compare it as those might give confused results.
// As Boolean( [] ) = true; Boolean( {} ) = true
// However Number( [] ) = 0; Number( {} ) = NaN

[] == false
// true

!![] == false
// false
```

Author: > Number( {} ) = NaN

It seems correct for me. It's not a number. But the rest seems :ok_hand:

```javascript
false === 'false'
// false

false == 'false'
// true

// 1 === '1'
// false

// 1 == '1'
// true
```

#### 2.4.6. Booleans, Truthies & Falsies
```javascript
// Booleans:
true, false

// Truthy:
"foo", 1

// Falsy:
"", 0, null, undefined, NaN, void 0
```

### 2.5. Practical Style

#### 2.5.1. Module pattern

```javascript
( function() {
	var Module = ( function() {
		var data = "secret";

		return {
			// This is some boolean property.
			bool: true,
			// Some string value.
			string: 'a string',
			// An array property.
			array: [ 1, 2, 3, 4 ],
			// An object property.
			object: {
				lang: 'en-Us'
			},

			getData: function() {
				// Get the current value of `data`.
				return data;
			},

			setData: function( value ) {
				this.value = value;
			}
		};
	} )();

	// Other things might happen here.

	// Expose our module to the global object.
	CKEDITOR.module = Module;
} )();
```
#### 2.5.2. Practical constructor

```javascript
( function() {

> 	var fooBar = CKEDITOR.tools.createClass( {

Reviewer: ```suggestion
	var FooBar = CKEDITOR.tools.createClass( {
```

We should use PascalCase for class names.

Reviewer: Please read also: https://github.com/ckeditor/ckeditor-idiomatic.js/pull/2#discussion_r370122285 bfore further comments.

Author: Agree for using PascalCase for constructors created with `createClass` as they have the same purpose as function constructors.

		$: function( foo ) {
			this.foo = foo;
		},
		
		proto: {
			getFoo: function() {
				return this.foo;
			},

			setFoo: function( val ) {
				this.foo = val;
			}
		}
	} );

	// Expose our constructor to the global object.

> 	CKEDITOR.fooBar = fooBar;

Reviewer: ```suggestion
	CKEDITOR.fooBar = FooBar;
```

Reviewer: Oh I see we don't use pascal case inside `CKEDITOR` namespace :(
However it is used internally for classes.
Maybe there should be reference that outside of `CKEDITOR` namespace we're using PascalCase for describing classes?
Like here: https://github.com/ckeditor/ckeditor4/blob/01b82d9d86eb1ebabf68934dcf71823f12bf7db8/plugins/autocomplete/plugin.js#L159

Author: Already covered with types naming notation.

} )();
```

#### 2.5.3. Prototype extension

```javascript
( function() {

> 	var fooBar = CKEDITOR.tools.extend( FooBar.prototype, {

Reviewer: ```suggestion
	CKEDITOR.tools.extend( FooBar.prototype, {
```

When you extend prototype you probably not expect to assign it to any variable.
alternative example could look like:

```javascript
( function() {
	var FooBar = CKEDITOR.tools.createClass( {
		// class definition
	} );

	// …	

	CKEDITOR.tools.extend( FooBar.prototype, {
		getFoo: function() {
			return this.foo;
		},

		setFoo: function( val ) {
			this.foo = val;
		}
	} );

	// Expose our constructor with extended prototype to the global object.
	CKEDITOR.fooBar = FooBar;
} )();
```

		getFoo: function() {
			return this.foo;

> 		};

Reviewer: ```suggestion
		},
```

		setFoo: function( val ) {
			this.foo = val;

> 		};

Reviewer: ```suggestion
		}
```

	} );

	// Expose our constructor to the global object.
	CKEDITOR.fooBar = fooBar;
} )();
```

### 2.6. Naming

You are not a human code compiler/compressor, so don't try to be one.

The following code is an example of egregious naming:

```javascript
function q(s) {
	return document.querySelectorAll(s);
}
var i,a=[],els=q("#foo");
for(i=0;i<els.length;i++){a.push(els[i]);}
```

Here's the same piece of logic, but with kinder, more thoughtful naming (and a readable structure):
```javascript

function query( selector ) {

>   return document.querySelectorAll( selector );

Reviewer: ```suggestion
	return document.querySelectorAll( selector );
```
Here should be tabulator to match other indentation.

}

var elements = [],
	matches = query( '#foo' ),
	length = matches.length;

for ( var i = 0; i < length; i++ ) {
	elements.push( matches[ i ] );
}
```

#### 2.6.1. Methods and functions

Methods and functions are **almost always** verbs/actions in `camelCase` notation:

```javascript
// Good
execute();
this.getNextNumber();

// Bad
this.editable();
this.status();
```

The exception for the above are get/set methods that are used very often, where the first parameter is the thing to get/set and the second is the optional value (set with it and get without):

```javascript
this.data( 'name' );  // get
this.data( 'name', 'John' );  // set

this.attr( 'title' );  // get
this.attr( 'title', 'Flying plane' );  // set
```

#### 2.6.2. Properties and variables

Properties and variables are **almost always** nouns in `camelCase` notation:

```javascript
var editor;
this.name;
this.editable;
```
Boolean properties and variables are **always** prefixed by an auxiliary verb:

```javascript
this.isDirty;
this.hasChildren;
this.isFake;
```

> Properties that require computation can use getters/setters and are named as nouns instead of having them as getXXX() and setXXX() methods when they are programmatically and conceptually simple:

Reviewer: ```suggestion
Properties that require computation can use getters/setters and are named as nouns instead of having them as `getXXX()` and `setXXX()` methods when they are programmatically and conceptually simple:
```

```javascript
// Good
this.status; // getter
editor.setReadOnly( true );
editor.readOnly;

// Bad
this.status = x; // setter
this.setReadOnly( true );
this.getReadOnly();

> ```

Reviewer: Those examples are really similar. I have problem with understanding why `editor.setReadOnly( true )` is correct however `this.setReadOnly( true )` is not. Maybe those examples should be presented in a different way. And there should be provided information that editor contains some exceptions to those rules, which are a legacy content and are not changed to preserve backward compatibility.

Author: Actually, we can just drop bad examples to simplify it and still keep original message.

#### 2.7.1. Event listeners

Keep `this` "local" to the class. This means &ndash; do not bind e.g. a function in `selection.js` file to editor instance and do not create a function which should be ran in editor context. This way you'll avoid confusion what `this` refers to.

```javascript
function Selection() {
	var editor = this.editor;

	// Bad:
	this.editor.on( 'setData', someCallback, editor );

	// Still bad:
	this.editor.on( 'setData', CKEDITOR.tools.bind( someCallback, editor ) );

	// Good:
	this.editor.on( 'setData', someCallback, this );

	// Good:
	this.getEditor().on( 'setData', function( evt ) {
		evt.editor.doSomething();
	} );
}

function someCallback() {
	// Should refer to Selection, not Editor.
	var editor = this.editor;
	// ...
}
```

#### 2.7.2. Use `thisArg` argument

Several prototype methods of ES 5.1 built-ins come with a special `thisArg` signature, which should be used whenever possible. CKEditor follows this practice too, so for example [`eventTarget.on`](http://docs.ckeditor.com/#!/api/CKEDITOR.event-method-on) also accepts context as an argument.

```javascript
var obj = { f: 'foo', b: 'bar', q: 'qux' };

Object.keys( obj ).forEach( function( key ) {

>   // |this| now refers to `obj`

Reviewer: ```suggestion
	// |this| now refers to `obj`
```
Here should be tab as in other examples.

>   console.log( this[ key ] );

Reviewer: ```suggestion
	console.log( this[ key ] );
```
Here should be tab as in other examples.

> # Code style guide for [CKEditor](https://github.com/ckeditor/ckeditor-dev)

Reviewer 2: ```suggestion
# Code style guide for [CKEditor 4](https://github.com/ckeditor/ckeditor4)
```

Author: Weirdly I can't commit that but :+1:

> # 1. Introduction

Reviewer 2: I'd add this heading, without deleting the main title. It would however involve moving the whole heading outline one level lower.

Author: So, in that case, the lowest heading would have 5 dots which is plenty (`1.1.1.1.1.`). I was trying to avoid such long nesting :thinking: Maybe we could go with just additional header at the top (so restoring previous `Code style guide for CKEditor` without introducing numbering for that? Not sure if it's correct (probably not) :thinking:

Author: Ok seems like I incorrectly understood your comment. It's about changing headers level only, so :+1:

> **This document is based on [Idiomatic.JS](https://github.com/rwldrn/idiomatic.js).** However, it contains many clarifications and rules specific for projects maintained by [CKSource](http://cksource.com/) like [CKEditor](https://github.com/ckeditor/ckeditor-dev).

Reviewer 3: ```suggestion
**This document is based on [Idiomatic.JS](https://github.com/rwldrn/idiomatic.js).** However, it contains many clarifications and rules specific for projects maintained by [CKSource](https://cksource.com/) like [CKEditor 4](https://github.com/ckeditor/ckeditor4).
```

> ### "Part of being a good steward to a successful project is realizing that writing code for yourself is a Bad Idea™. If thousands of people are using your code, then write your code for maximum clarity, not your personal preference of how to get clever within the spec."
>_Idan_ _Gazit_

> ## 1.2. Table of contents

Reviewer 2: Shouldn't it be at the top, before even introduction?

> ## 1.2. Table of contents

Reviewer 3: I think numbers would be better in table of contents than bullets.

* [Code style guide](#2-code-style-guide)
	* [Whitespace](#21-whitespace)
	* [Beautiful Syntax](#22-beautiful-syntax)
	* [Type Checking (Courtesy jQuery Core Style Guidelines)](#23-type-checking)
	* [Conditional Evaluation](#24-conditional-evaluation)
	* [Practical Style](#25-practical-style)
	* [Naming](#26-naming)
	* [Faces of `this`](#27-faces-of-this)
	* [Misc](#28-misc)
	* [Native & Host Objects](#29-native--host-objects)
	* [Comments](#210-comments)
* [Tests](#3-tests)
	* [Manual tests](#31-manual-tests)

# 2. Code style guide

> ## 2.1 Whitespace

Reviewer: What about BOM in files?
It looks like a tool for updating languages in plugins re-implement a BOM at the beginning of a file. Maybe it could be defined here as well?
Here an example when it is included:
```
git diff f4bb06c bd5101a -- plugins/language/plugin.js
```
https://github.com/ckeditor/ckeditor4/compare/f4bb06c..bd5101a#diff-beff19e4c26b0a4ac52a409f337b0258

Author: BOM is just incorrect, not sure if we should add docs for that.

- Always use tabs. Never use spaces for indentation (for both - code and comments).

> - Tab is 4 spaces.

Reviewer 2: I do not see why defining tab width is necessary. That's the whole point of tabs: you don't need to define its width.

Reviewer 4: Agree with @Comandeer here, tab is tab and everyone can set it's own width but at the end it still one tab character.

Author: Overall, yes, but when you use e.g. vim, you can configure tab as 2 spaces. Although, I agree that it's no longer tab than, so it's redundant information (actually, misleading, because if you count tab in spaces, it's no longer tab).

> - Do not leave trailing spaces (**note:** empty lines should not contain any spaces).

Reviewer 3: Isn't this note redundant? Spaces in empty lines are still trailing I guess?

Author: Overall yeah, but I would go for small redundancy for the sake of clearness.

- Always use LF line endings. Never use CRLF.
- If your editor supports it, always work with the "show invisibles" setting turned on. The benefits of this practice are:
  - Enforced consistency
  - Eliminating end of line whitespace
  - Eliminating blank line whitespace
  - Commits and diffs that are easier to read

> Related settings for Sublime users:

Reviewer 2: IMO it's better to include more standard format, [EditorConfig](https://editorconfig.org/). ST is not as popular as it was when this guide was created.

Reviewer 3: I'd add also a note that each file should end with an empty line and that it can be also set in most editors config.

Author: @Dumluregn good point.

	{
		// Defaults:
		"tab_size": 4,
		"translate_tabs_to_spaces": false,

		// Additional settings:
		"default_line_ending": "unix",
		"trim_trailing_white_space_on_save": true
	}

## 2.2. Beautiful syntax

### 2.2.1. Parens, Braces, Linebreaks

> #### 2.2.1.1. Use whitespace to promote readability

Reviewer 2: Probably we should describe in details rules for whitespace, like "always insert whitespace before and after `(` etc.

Reviewer 4: Isn't the code sample below enough to understand it? 🤔

Reviewer 3: I also think that with a 2.2.1.1 header it's rather self-explanatory.

Reviewer 3: Unless you mean that we should add more examples, e.g. not adding whitespace before function arguments - I see it isn't covered anywhere, so :+1: from me.

Author: Not sure if it's worth adding more examples for such simple rule which always generates some noise :thinking:

Author: We decided to skip additional examples.

```javascript
if ( condition ) {
	// statements
}

if ( true ) {
  // statements
} else {
  // statements

> }

Reviewer 4: Something with indentation is broken in this code sample.

Reviewer 3: What do you mean? Looks ok for me 🤔

if ( true ) {
	// statements
}
// Comments regarding what "else" means and does.
else {
	// statements
}

while ( condition ) {
	// statements
}

for ( var i = 0; i < 100; i++ ) {
	// statements
}

try {
	// statements
} catch ( e ) {
	// statements
}

```

#### 2.2.1.2. Always use brackets even for a single line statements

```javascript
// Bad
if ( true ) // statement

// Good
if ( true ) {
	// statement
}

// Bad
if ( false ) // statement
else // statement

// Good
if ( false ) {
	// statement
} else {
	// statement
}

> ```

Reviewer 4: Maybe just show "Good" examples here without polluting it with too much code?

Reviewer 3: I think that including "Bad" ones helps noticing what exactly is unwelcome.

### 2.2.2. Assignments, Declarations, Functions

> #### 2.2.2.1. Declaring variables

Reviewer 3: I would change this to sound more like a rule (like the other headers), e.g. "Don't duplicate `var` statement".

Author: Actually, this whole header seems redundant.

```javascript
var foo = 'bar',
	num = 1,
	undef;

// Literal notations:
var array = [],
	object = {};
```

#### 2.2.2.2. Single line per declaration

> Declaring one variable per line improves debugging experience. Multiple variables declared in the same line won't allow you to place breakpoint at the definition computation. The same goes to object literals.

Reviewer 3: It also makes it easier to find a particular variable declaration.

```javascript
// Bad
var foo = 'foo', bar = 'bar',
	object = {
		foo: 'foo', bar = 'bar' // Also bad
	};

// Good
var foo = 'foo',
	bar = 'bar',
	object = {
		foo: 'foo',
		bar: 'bar'
	};

```

#### 2.2.2.3. One variable per scope

> Using only one variable per scope (function) promotes readability and keeps your declaration list free of clutter (also saves a few keystrokes). However, don't overuse it. If a scope have multiple lines of code, it may be more readable to declare variable later when it's defined or used than expect developer to scroll up to declaration list instead.

Reviewer 2: I'm for slightly modifying this rule: declare variables at the top of current scope. Functions shouldn't be long enough for the need of declaring some variables later. If it is so long, it should be divided into several smaller functions with their own variable declarations at the top.

Reviewer 4: I mostly agree, but there are some cases when it decreases readability IMHO:

```js
function( arg1, arg2 ) {
    var result = [];

    if ( !arg1 ) {
        return;
    }

    // ...rest of function body
}
```

vs

```js
function( arg1, arg2 ) {
    if ( !arg1 ) {
        return;
    }

    var result = [];
    // ...rest of function body
}
```

Reviewer 3: Besides readability, such early returns can also upgrade performance a bit if some variables will only be declared (by hoisting) and not initialized. But OTOH it's quite risky, what if we want to have 3 early returns and some of them checks vars initialized in function scope? Trying to overoptimize this may lead to shattering the code too much. So I guess some common sense is necessary in such cases.

Author: I see all 3 comments correct :+1:

```javascript
// Bad.
function foo() {
	// 2 statements

	var bar = '';

	// 2 statements

	var foo = '';

	// 10 statements

	var bom = '';

	// some statements
}

// Good
function foo() {
	var bar = '',
		foo = '';

	// 10 statements

	var bom = '';   // Assuming that bom isn't needed before, there's no sense
					// in grouping it with the first group.

	// some statements
}
```

#### 2.2.2.4. Function declaration

> Prefer function declaration instead of [function expression](#2225-function-expression) to avoid issues with executing function before it has been defined.

Reviewer 2: Probably we should also add rules about declaring functions after the main flow to promote greater readability.

Reviewer 4: It's more about good coding practises then code style so we could probably add more like Single Responsibility etc. but I'm not sure code style guide it's the right place 🤔

Author: @Comandeer it has been covered with `#### 2.2.3.2. Helper functions`.
@f1ames I would say it's more about code style, if I understand correctly, @Comandeer wrote only about function placement. I would also avoid adding too much information especially about correct coding, or we will end up with a book :thinking:

```javascript
function foo( arg1, argN ) {

}

// Usage.
foo( arg1, argN );

// Real life example.
function square( number ) {
	return number * number;
}

// Usage.
square( 10 );
```

> #### 2.2.2.5. Function expression

Reviewer 4: Maybe just skip it since it's not recommended and it's mentioned in the previous section too?

Prefer [function declaration](#2224-function-declaration) when possible.

```javascript

>     var square = function( number ) {

Reviewer: I would add here some comment like:
```
// not recommended
```
or
```
// avoid such construciton
```

Reviewer 3: Or move code to beginning 2.2.2.4 listing and mark it as "Bad". That will make it clear how it should look like instead.

Author: > Prefer [function declaration](#2224-function-declaration) when possible.

Don't you see it enough? :thinking:

        return number * number;
    };
```

> #### 2.2.2.6. Continuation-Passing style

Reviewer 4: If we are on async/ca;;back code should we mention something about promises?

> If your function should be executed asynchorously or requires additional callback (e.g. filter function) as an function argument, pass it as the last argument in function signature.

Reviewer 3: Fix typo: `asynchorously ` -> `asynchronously`.

```javascript
// Declaration.
function square( number, callback ) {
	callback( number * number );
}

// Usage.
square( 10, function( square ) {
	// callback statements
} );
```

> #### 2.2.2.7 Constructor declaration

Reviewer 2: Shouldn't we promote usage of [`CKEDITOR.tools.createClass`](https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_tools.html#method-createClass)?

Reviewer: 👍 from me. It unifies a lot.
For example, when you check how `BalloonToolbar` class, my first thought here was: WTF?!. There is a completely different pattern which is not coherent with other places where `CKEDITOR.tools.createClass` is used.

```javascript
function FooBar( options ) {
	this.options = options;
}

// Usage.
var fooBar = new FooBar( {
	a: 'alpha'
} );

fooBar.options; // -> { a: 'alpha' }
```

### 2.2.3. Code structure

#### 2.2.3.1. Public API

Exposing public API should have much higher priority in file structure than private members. It improves API discoverability and hides private members.

```javascript
// Simple module.
CKEDITOR.module1 = {
	pubFoo: function() {
		this.privFoo();
	},

	pubBar: function() {
		this.privBar();
	},

	privFoo: function() {
		// ...
	},

	privBar: function() {
		// ...
	}
};

// Using module pattern.
var module = ( function() {
	// Exposing public API.
	return {
		foo: foo,
		bar: bar
	}

	function foo() {
		bar();
	}

	// Private helpers.
	function bar() {
		// ...
	}
} )();
```

#### 2.2.3.2. Helper functions

> When extracting helper function, move it right after the first usage inside the nearest scope which won't affect code performance and memory usage. In most cases, you want to keep helper functions on the same scope level as a public one unless they are used somewhere else in the same file.

Reviewer 4: If you have two public classes/modules in the same file I would be for throwing are helpers to the end of the file. It's consistent to a previous point:

> Exposing public API should have much higher priority in file structure than private members.

So all public stuff goes first and then private helpers. According to what is now propose you will have public - helpers - public - helpers mix. WDYT?

```javascript
CKEDITOR.module1 = {
	doSomething: function() {
		// Good
		foo();

> 		// Bad thus the function object will be created every time when doSomething function executed.

Reviewer: Shouldn't this comment be attached to `function bar()` line?

		bar();

		function bar() {
			// ...
		}
	}
};

function foo() {
	baz();
	quix();
}

function baz() {
	// ...	
}

function quix() {
	// ...
}

CKEDITOR.module2 = {
	doSomething: function() {
		// Note that despite this function is used in another module later,
		// it has been correctly declared as close as possible to the first usage.

> 		foo();

Reviewer: I wonder if this approach is good one. Having @Comandeer comment on mind:
https://github.com/ckeditor/ckeditor-idiomatic.js/pull/2/files#r363742631
It seems that better would be moving all function (foo, baz, quix) declaration below it. Or some better description of how such cases should be treat.

Author: It's more like a separate example actually, but yeah, let's just move it and treat as a whole.

Author: Not sure here, keeping privates close makes them easier to find :thinking:

```javascript
CKEDITOR.module1 = {
	foo: function() {
		this.bar();
	},

	bar: function() {
		// ...
	},

	baz: function() {
		// ...
	}
};
```

#### 2.2.3.3. Configuration members

Configuration members exposed by `CKEDITOR.config` object should be moved at the end of the code file. It's adapted practice in CKEditor code base which improves configuration options discoverability.

```javascript
CKEDITOR.plugins.add( 'myplugin', definition );

// ... some code goes here ...
// ... more lines of code ...
// ... just don't wake up Cthulhu ...

CKEDITOR.config.foo = 'foo';
CKEDITOR.config.bar = 'bar';
```

### 2.2.4. Quotes

Always use single quotes in JavaScript, but double quotes in HTML.

## 2.3. Type checking

### 2.3.1. Actual Types

String:

```javascript
typeof variable === 'string'
```

Number:

```javascript
typeof variable === 'number'
```

Boolean:

```javascript
typeof variable === 'boolean'
```

Object:

```javascript
typeof variable === 'object'
```

Array:

```javascript
// In CKEditor:
CKEDITOR.tools.isArray( arrayLikeObject )

> // In other cases:

Reviewer 3: What other cases you have in mind?

Author: Like test files where we don't have CKE or some docs samples. Not very common, but possible.

Array.isArray( arrayLikeObject )
(wherever possible)
```

Node:

```javascript
// In CKEditor:

> elem.type == CKEDITOR.NODE_ELEMENT

Reviewer 2: ```suggestion
elem.type === CKEDITOR.NODE_ELEMENT
```

// In other cases:

> elem.nodeType == 1

Reviewer 2: ```suggestion
elem.nodeType === Node.ELEMENT_NODE
```

```

null:

```javascript
variable === null
```

undefined:

```javascript

> variable === undefined

Reviewer 2: I'm not sure about it. Personally I use `typeof variable === 'undefined'`. It's more safe, especially in code that is not always in strict mode.

Reviewer: I believe that using `typeof` for `undefined` is more confusing. I don't see how all our checks with undefined will now could be replaced with `typeof`.
I prefer readability with `===` over safe option proposed here.

```

> #### 2.3.2. Coerced Types

Reviewer: I'm missing `!!` for Boolean coercion. It's quite common practice to do so and I feel should be included here as well.

Reviewer 3: Maybe move comments showing results inline or before expressions? Would be more readable for me.

Author: I would say it's bad practice, but yeah, pretty common. We should rather use `Boolean( x )` instead for explicitness. But let's discuss it because AFAIK I'm on losing side here (with the team).

number + '';
// '1'
// ...prefer explicit parsing to string
String( number );
// '1'

string;
// '1'

// Insead of...
+string;
// 1
// ...prefer explicit parsing to number
Number( string );
// 1

bool;
// false

// Insead of...
+bool;
// 0
// ...prefer explicit parsing to number
Number( bool );
// 0

// Insead of...
bool + '';
// 'false'
// ...prefer explicit parsing to string
String( bool );
```

### 2.3.3. Coersion practices to avoid

Avoid using unpopular coersion operators due to lacking readability. Most of them have explicit versions, which are much more readable.

#### 2.3.3.1. Bitwise NOT operator

Instead of bitwise NOT operator `~` prefer explicit aproach by comparing returned value.

```javascript
var array = [ 'a', 'b', 'c' ];

// Bad
if ( ~array.indexOf( 'a' ) ) {
	// ...
}

// Good
if ( array.indexOf( 'a' ) >= 0 ) {
  // ...
}
```

#### 2.3.3.2. Double bitwise NOT operator
Instead of double bitwise NOT operator `~~` resulting in numerical substitution, use `Math.floor` or `parseInt` methods.

```javascript
var num = 2.5;

// Bad
~~num;

// Good
parseInt( num, 10 );

// Good
Math.floor( num );
```

## 2.4. Conditional Evaluation

> ### 2.4.1. Boolean array evaluation

Reviewer 2: This seems to be in contradiction with previous points. Treating length as a boolean value decreases readability and causes some cognitive overhead. I'm for explicit comparison to `0`.

Reviewer: I would only force comparison to `0` when `array` variable might not be an Array. So `length` property might be `undefined` what will result with always `falsy` condition. For other cases, I'm fine with treating length as boolean.

Author: It would be sad to not take advance of coercion, at least partially with the most popular use cases.

```javascript
// When only evaluating that an array has length,
// instead of this:
if ( array.length > 0 ) ...

// ...evaluate truthiness, like this:
if ( array.length ) ...

// When only evaluating that an array is empty,
// instead of this:
if ( array.length === 0 ) ...

// ...evaluate truthiness, like this:
if ( !array.length ) ...
```

> ### 2.4.2. Boolean string evaluation

Reviewer 2: In this case, I'm also for more explicit check: `string.length === 0` or `string !== ''`.

Author: It may be hard to introduce such code style with current codebase :thinking:

Reviewer 4: Let's be explicit in such cases.

Author: I believe we discussed beeing implicit here to avoid too extensive changes to our codebase, but if someone disagrees, please tell.

```javascript
// When only evaluating that a string is not empty,
// instead of this:
if ( string !== '' ) ...

// ...evaluate truthiness, like this:
if ( string ) ...

// When only evaluating that a string _is_ empty,
// instead of this:
if ( string === '' ) ...

// ...evaluate falsyness, like this:
if ( !string ) ...
```

### 2.4.3. Boolean evaluation

```javascript
// When only evaluating that a reference is true, instead of this:
if ( foo === true ) ...

// ...evaluate like you mean it, take advantage of built in capabilities:
if ( foo ) ...

// When evaluating that a reference is false, instead of this:
if ( foo === false ) ...

// ...use negation to coerce a true evaluation
if ( !foo ) ...

// ...Be careful, this will also match: 0, '', null, undefined, NaN. If you MUST test for a boolean false, then use:
if ( foo === false ) ...
```

### 2.4.4. Boolean ref evaluation

```javascript
// When only evaluating a ref that might be null or undefined,
// but NOT false, '' or 0, instead of this:
if ( foo === null || foo === undefined ) ...

> // ...take advantage of == type coercion, like this:

Reviewer: I don't like this approach.
Generally, I would avoid at all cost situation that some variable might be `undefined` or `null` and then might be processed differently. And this approach hides such situation even further.

Reviewer 3: Yeah, I'm also not sure about such approach, I think it's a bit too clever and may be misleading.

Author: Not sure here :thinking: But yeah, it's a bit not readable. I'm missing some nice helper like `CKEDITOR.isNullOrUndefined` honestly.

Reviewer 4: Let's use `if ( foo === null || foo === undefined )`.

if ( foo == null ) ...

// Remember, using == will match a `null` to BOTH `null` and `undefined` but not `false`, '' or 0
null == undefined
```

ALWAYS evaluate for the best, most accurate result - the above is a guideline, not a dogma.

### 2.4.5. Loose equality coersion 

> Use loose equality operator to simplify your code when you want to leverage type coercion. Remember that strict equality comparator `===` is checking if both the type and the value you are comparing are the same. In a constract, loose equality operator `==` will try to do type coersion which may be useful in some cases when comparing values where types are less revelant than values.

Reviewer 3: I'd add a reference to a cheatsheet here, like [this one](https://dorey.github.io/JavaScript-Equality-Table/).

Author: I'm a bit skeptical about adding 3rd party references which are not official docs :thinking: Such links like to go out of date or provide incorrect information. Can you share some official docs instead?

```javascript
false === 'false'
// false

false == 'false'
// true

> // 1 === '1'

Reviewer 3: I guess this shouldn't be commented?

// false

> // 1 == '1'

Reviewer 3: This also shouldn't be put in a comment.

// true
```

### 2.4.6. Booleans, Truthies & Falsies
```javascript
// Booleans:
true, false

// Truthy:
"foo", 1

// Falsy:
"", 0, null, undefined, NaN, void 0
```

## 2.5. Practical Style

> ### 2.5.1. Module pattern

Reviewer 2: Do we even use such pattern?

Reviewer 4: ☝️

Author: https://github.com/ckeditor/ckeditor4/blob/8794f9fdd9bcd4cbbd75530dc94e3ca3148731d3/tests/plugins/easyimage/_helpers/tools.js#L1-L30

```javascript
( function() {
	var Module = ( function() {
		var data = "secret";

		return {
			// This is some boolean property.
			bool: true,
			// Some string value.
			string: 'a string',
			// An array property.
			array: [ 1, 2, 3, 4 ],
			// An object property.
			object: {
				lang: 'en-Us'
			},

			getData: function() {
				// Get the current value of `data`.
				return data;
			},

			setData: function( value ) {
				// Set the value of `data` and return it.
				return ( data = value );
			}
		};
	} )();

	// Other things might happen here.

	// Expose our module to the global object.
	CKEDITOR.module = Module;
} )();
```

> ### 2.5.2. Practical constructor

Reviewer 2: Probably we should have also some example with `CKEDITOR.tools.createClass`.

```javascript
( function() {
	function Ctor( foo ) {
		this.foo = foo;

		return this;
	}

	CKEDITOR.tools.extend( Ctor.prototype, {
		getFoo: function() {
			return this.foo;
		};

		setFoo: function( val ) {
			return ( this.foo = val );
		};
	} );

	// Expose our constructor to the global object.

> 	CKEDITOR.ctor = Ctor;

Reviewer 4: Looking into it wondering "WTH is Ctor". I see `Constructor` is not a reserved JS keyword (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Lexical_grammar#Keywords) so maybe use the full name.

} )();
```

> ## 2.6. Naming

Reviewer: I would also extend this section with a description what words used to name different things.
For example, use verbs to describe the function, or nouns for variables etc.
In a similar way as it is done for CKE5 here
https://github.com/ckeditor/ckeditor5-design/wiki/Code-Style-Naming-Guidelines

You are not a human code compiler/compressor, so don't try to be one.

> The following code is an example of egregious naming:

Reviewer 2: Still totally readable to me :smile:

Reviewer 4: > You are not a human code compiler/compressor

We all know ☝️ is an incorrect assumption in your case @Comandeer 😄

```javascript
function q(s) {
	return document.querySelectorAll(s);
}
var i,a=[],els=q("#foo");
for(i=0;i<els.length;i++){a.push(els[i]);}
```

Here's the same piece of logic, but with kinder, more thoughtful naming (and a readable structure):
```javascript

function query( selector ) {
  return document.querySelectorAll( selector );
}

var id = 0,
	elements = [],
	matches = query( '#foo' ),
	length = matches.length;

> for ( ; id < length; id++ ) {

Reviewer: I don't like that this "better" example skips 1st expression in `for` loop.

Author: Do you find out skipping 1st statement a bad practice? IMO it's totally fine, but I will update it for you :cry:

Author: Yeah, overall the example is weird due to id definition earlier, so I will fix it. But there is nothing wrong with skipping the first statement if needed.

	elements.push( matches[ id ] );
}
```

### 2.6.1. Notation

#### 2.6.1.1. Variables, functions, objects

Use `camelCase` notation for naming:

* variables

```javascript
var camelCaseVariableName = 'foo!';
```

* functions
```javascript
function camelCaseFunctionName() {}
```

* methods
```javascript
object.prototype.camelCaseMethodName = function() {};
```

* objects
```javascript
var camelCaseObjectName = {
	foo: 'foo',
	bar: function() {}
};
```

* instances
```javascript
var camelCaseInstance = new Object();
```

* public members
```javascript
CKEDITOR.sth.camelCase = sth;
```

#### 2.6.1.2. Types

Use `PascalCase` notation for naming constructor function:

```javascript
function PascalCase () {}

PascalCase.prototype = {
	foo: function() {},
	bar: function() {}
};
```

> #### 2.6.1.3. Constants

Reviewer 3: Agree about convention, but we don't really use `const` variables, so I don't think it's necessary (especially with `var` statement).

Author: It's not about specific `const` language keyword but much older constants convention.

Use `UNDERSCORE_SNAKE_CASE` for naming constant variables:

```javascript
// Constant variables.
var CONSTANT_VARIABLE = 999;

// Constant enum.
CKEDITOR.sth.CONSTANT_ENUM = 1;
```

#### 2.6.1.4. Private members

Only use underscore `_` name as an object member with private properties:

```javascript
var object = {
	// Public member.
	foo: {},
	// Also public member.
	bar: {},
	// Private member and everything contained inside this object is also private.
	_: {}
}
```

> Avoid using underscore `_` with private variables, methods and functions. Note that all undocumented members are private by default, so there is no need beeing explicit here.

Reviewer 2: We have plenty of private methods with `_` at the beginning. I'd even risk saying that our current convention is totally different: we add underscore to indicate that some property is private.

Reviewer 4: 🤔 I tried to avoid it too, so I'm not sure in this case. I'm fine with convention that not documented methods are private.

Author: Treating everything private by default is common and pretty good practice. Underscores make sense in languages where you cannot easily hide references like in Python, but in JS we can hide almost everything in scopes. So, what's the point?

> 'd even risk saying that our current convention is totally different

Not sure, but our codebase says different story :thinking:

var object = {
	/*
	 * @private
	 */
	 foo: {},

	/*
	 * Some nice description here.
	 * @property {Object}
	 */
	 bar: {}
};
```

#### 2.6.1.5. Configuration members

If configuration member is a part of core codebase, go with `camelCaseNaming` notation.

```javascript
CKEDITOR.config.fooBarBaz = true;
```

If configuration member is a part of [plugin](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_plugins.html) functionality, use `pluginName_optionName` pattern consisting of mixed `camelCase` plugin and option naming separated with underscore `_`.

```javascript
CKEDITOR.config.myPlugin_myProperty = true;
```

## 2.7. Faces of `this`

### 2.7.1. Event listeners

Keep `this` "local" to the class. This means &ndash; do not bind e.g. a function in `selection.js` file to editor instance and do not create a function which should be ran in editor context. This way you'll avoid confusion what `this` refers to.

```javascript
function Selection() {
	var editor = this.editor;

	// Bad:
	editor.on( 'setData', someCallback, editor );

	// Still bad:
	editor.on( 'setData', CKEDITOR.tools.bind( someCallback, editor ) );

	// Good:

> 	editor.on( 'setData', someCallback, this );

Reviewer 4: A little confusing because in "bad" examples you pass/bind editor and in "good" one you pass different object (`this` which needs `this.editor`), this seem inconsistent and a little confusing about the intention of this example.

	// Good:
	this.getEditor().on( 'setData', function( evt ) {
		evt.editor.doSomething();
	} );

	// Good too:
	var that = this;
	editor.on( 'setData', function( evt ) {
		editor.setCurrentSelection( that );

> 	} );

Reviewer 4: You can do

```js
	editor.on( 'setData', function( evt ) {
		editor.setCurrentSelection( this );
	}, this );
```

so why bother with `that`?

Reviewer 4: And the next section is exactly about that.

}

function someCallback() {
	// Should refers to Selection, not Editor.
	this;
}
```

### 2.7.2. Use `thisArg` argument

Several prototype methods of ES 5.1 built-ins come with a special `thisArg` signature, which should be used whenever possible. CKEditor follows this practice too, so for example [`eventTarget.on`](http://docs.ckeditor.com/#!/api/CKEDITOR.event-method-on) also accepts context as an argument.

```javascript
var obj = { f: 'foo', b: 'bar', q: 'qux' };

Object.keys( obj ).forEach( function( key ) {

  // |this| now refers to `obj`

  console.log( this[ key ] );

}, obj ); // <-- the last arg is `thisArg`

// Prints...

// 'foo'
// 'bar'
// 'qux'
```

> `thisArg` can be used with `Array.prototype.every`, `Array.prototype.forEach`, `Array.prototype.some`, `Array.prototype.map`, `Array.prototype.filter`.

Reviewer 3: Keep in mind that all of those are not working on IE8 and we have polyfills for them, so maybe refer to them instead.

Reviewer 4: Let's use examples with `CKEDITOR.tools`.

### 2.7.3. Binding `this` to the function scope

Beyond the generally well known use cases of `call` and `apply`, always prefer `.bind( this ) / CKEDITOR.tools.bind( fn, this )` for creating `BoundFunction` definitions for later invocation.

```javascript
function Device( opts ) {
	stream.read( opts.path, CKEDITOR.tools.bind( function( value ) {
		this.value = value;
	}, this ) );
}
```

Only resort to aliasing when no preferable option is available.

```javascript
function Device( opts ) {
	var self = this;
	stream.read( opts.path, function( value ) {
		self.value = value;
	} );
}
```

## 2.8. Misc

This section will serve to illustrate ideas and concepts that should not be considered dogma, but instead exists to encourage questioning practices in an attempt to find better ways to do common JavaScript programming tasks.

### 2.8.1. Early returns

Early returns promote code readability with negligible performance difference.
```javascript
// Bad:
function returnLate( foo ) {
	var ret;

	if ( foo ) {
		ret = 'foo';
	} else {
		ret = 'quux';
	}
	return ret;

> }

Reviewer 4: This example is unrealistic (I know it was simplified to be easy to understand) but looking on it, you think "why not go with a one-liner `return foo ? 'foo' : 'quux'`;". And yes, it promotes bad code as a side effect.

Reviewer: An example from our code:
https://github.com/ckeditor/ckeditor4/blob/294ea3333aa3d2b796f27ca8503ac7c1a658d50a/plugins/tableselection/plugin.js#L862-L890

// Good:
function returnEarly( foo ) {
	if ( foo ) {
		return 'foo';
	}
	return 'quux';
}

// Bad:
function doSthComplex( foo, bar ) {
	if ( foo ) {
		// do something

		if ( bar ) {
			// do something
		} else {
			// do something else
		}
	}
}

// Good:
function doSthComplex( foo, bar ) {
	if ( !foo )

> 		return;

Reviewer 2: ```suggestion
if ( !foo ) {
	return;
}
```

	// do something

	if ( !bar ) {
		// do something else
		return;
	}

	// do something
}
```

### 2.8.2. Objects minify poorly, closures very well

Avoid excessing usage of objects, when functional programming may be used as well. Properties and method names unlike private variables and function names cannot be shortened by minifier.

> You can play with [online Google Closure Compiler](http://closure-compiler.appspot.com/home) to test different approaches. Make sure to check the gzipped size which matters the most.

Reviewer 2: I'm not so sure about considering only size. We should also consider execution time, which is not 100% bound to the size of downloadable content.

Author: > Avoid excessing usage of objects, when functional programming may be used as well

But here you are writing about implementation correctness, not style guide :thinking:

Reviewer 4: 💣

## 2.9. Native & Host Objects

The basic principle here is:

> 	Don't do stupid shit and everything will be ok.

Reviewer 2: But extending objects we don't own is stupid shit ¯\_(ツ)_/¯

Author: Like polyfills? :thinking:

Reviewer 4: We should not extend native objects.

To reinforce this concept, please watch the following presentation:
	

> [“Everything is Permitted: Extending Built-ins” by Andrew Dupont (JSConf2011, Portland, Oregon)](https://www.youtube.com/watch?v=xL3xCO7CLNM)

Reviewer 4: 💣

**But** CKEditor is a widget - an application inside someone's system &ndash; so... do not touch anything except
`CKEDITOR.*` namespace.

> ## 2.10. Comments

Reviewer: I'm not sure where to put this. Maybe better would be the section about conditions. However, there should be explicit written that comment to a given if should be before it not inside it.

Reviewer 3: Please add info about ticket referencing convention.

- Single line above the code that is subject (except for long expressions like conditions).
- Comment is a sentence - start it with capital leter, end with period.
- Use multiline comments only for API documentation.
- Check the [JSDuck documentation guide](http://dev.ckeditor.com/wiki/CodeDocumentation).

> - You can use JSDuck documentation style for private functions, but then use single line comments.

Reviewer 2: Actually we often use multiline, but non-JSDocs comments:

```javascript
/*
 * Some docs for private functions.
 */
```

Author: I believe it's more for helper functions where we use `//` instead to avoid producing incorrect JSDocs :thinking: 

```
// Some docs for private functions.
// @param {Object} x
// @param {Object} y
// returns {Number}
```

Reviewer: Please consider comments like this one:
https://github.com/ckeditor/ckeditor4/blob/294ea3333aa3d2b796f27ca8503ac7c1a658d50a/plugins/basicstyles/plugin.js#L98-L102

Author: @msamsel it rather seems like this example is incorrect because array could be split across multiple lines instead of using multiline comment :thinking: WDYT?

Reviewer: yeah maybe it's not perfect then :D

- Justify end line comments with tabs:

```javascript
if ( some != really.complex &&      // Comment line 1.
	condition || with &&            // Comment line 2.
	( multiple == !lines ) )        // Comment line 3.

// Of course it's better to avoid this spaghetti at all.
```

# 3. Tests

## 3.1. Manual tests

> Manual tests reproduction steps should be structured in one, ubiquitous way, to unify test files, so they are easy to read and modify. Manual test file should consist only of a single test suite. Also, avoid putting many test cases in the same file, extract common logic instead using helper methods or template files.

Reviewer 2: I'd be careful with the rule of one manual test per file. It can cause the creation of tons of files.

Reviewer: Or it might favours the long test wrapped as a single one.

Author: But improves the testing phase. You can still create template file, but I'm for discussing this point.
